runtime/autoload/man.vim

```diff
@@ -1,137 +1,264 @@
+function! man#pop_tag() abort
+  if !empty(s:tag_stack)
+    let tag = remove(s:tag_stack, -1)
+    execute tag['buf'].'b'
+    call cursor(tag['lnum'], tag['col'])
   endif
+endfunction
 
-  exec 'let s:man_tag_buf_'.s:man_tag_depth.' = '.bufnr('%')
-  exec 'let s:man_tag_lin_'.s:man_tag_depth.' = '.line('.')
-  exec 'let s:man_tag_col_'.s:man_tag_depth.' = '.col('.')
-  let s:man_tag_depth = s:man_tag_depth + 1
-
-  let editcmd = 'edit'
-  " Use an existing 'man' window, else open a new one.
-  if &filetype !=# 'man'
-    let thiswin = winnr()
-    wincmd b
-    if winnr() > 1
-      exec thiswin . 'wincmd w'
-      while 1
-        if &filetype ==# 'man'
-          break
-        endif
-        wincmd w
-        if thiswin == winnr()
-          break
-        endif
-      endwhile
+" find the closest man window above/left
+function! s:find_man() abort
+  if &filetype ==# 'man'
+    return 1
+  endif
+  let thiswin = winnr()
+  while 1
+    wincmd w
+    if &filetype ==# 'man'
+      return 1
+    elseif thiswin == winnr()
+      return 0
     endif
+  endwhile
+endfunction
 
-    if &filetype !=# 'man'
-      let editcmd = 'tabnew'
-    endif
+let s:ref_char = '\%([^()[:space:]]\)'
+
+" returns the regex for a valid character in a manpage reference
+function! man#ref_char() abort
+  return s:ref_char
+endfunction
+
+let s:ref_chars = s:ref_char.'\+'
+
+" parses the page and sect out of 'page(sect)'
+function! s:parse_page_and_sect_fpage(fpage) abort
+  if a:fpage[0] ==# '-'
+    " to see why, disable this condition and try
+    " :echo systemlist('man 2>/dev/null -w -mandoc')
+    return ['', '']
+  elseif a:fpage =~? s:ref_chars.'([0-9nx][a-z]*)'
+    let ret = split(a:fpage, '(')
+    let iret = split(ret[1], ')')
+    return [ret[0], tolower(iret[0])]
   endif
+  return [matchstr(a:fpage, s:ref_chars), '']
+endfunction
 
-  silent exec editcmd.' man://'.page.(empty(sect)?'':'('.sect.')')
+" parses the page and sect out of 'path/page.sect'
+function! s:parse_page_and_sect_path(path) abort
+  if a:path =~# '\.\%([glx]z\|bz2\|lzma\|Z\)$' " valid extenions
+    let tail = fnamemodify(fnamemodify(a:path, ':r'), ':t')
+  elseif a:path =~# '\.\%(pdf\|in\)$' " invalid extensions
+    return ['', '']
+  else
+    let tail = fnamemodify(a:path, ':t')
+  endif
+  let page = matchstr(tail, '^.\+\ze\.[^.]\+$')
+  let sect = matchstr(tail, '\.\zs[^.]\+$')
+  return [page, sect]
+endfunction
 
+function! s:read_page(sect, page)
+  let bname = 'man://'.a:page.(len(a:sect)?'('.a:sect.')':'')
+  if s:find_man()
+    if bufexists(bname)
+      silent execute 'buf '.bufnr(bname)
+      return
+    endif
+    execute 'edit '.bname
+  elseif bufexists(bname)
+    execute 'split '.bname
+    call man#set_window_local_options()
+    return
+  endif
+  execute 'split '.bname
   setlocal modifiable
-  silent keepjumps norm! 1G"_dG
-  if empty($MANWIDTH)
+  setlocal noreadonly
+  if empty($MANWIDTH) || $MANWIDTH > winwidth(0)
```

Comment:
This will partially regress https://github.com/neovim/neovim/pull/4144. The reason for respecting MANWIDTH even if it is wider than the current window is that the current window might only be temporarily narrow. Since we don't reflow, we should just assume the user knows what they're doing and _actually respect_ MANWIDTH. (i.e., undo the change on this line)